R/denom.R

```diff
@@ -164,8 +164,35 @@ get_denom_total <- function(.data, denoms_by, denoms_df,
     .data$total <- ifelse(nrow(sums) > 0, sum(sums[["n"]], na.rm = TRUE), 0)
     # distinct_n is present for all count layers, but not shift layers, so
     # dont' do this for shift layers
-    if ("distinct_n" %in% names(sums))
-      .data$distinct_total <- ifelse(nrow(sums) > 0, sums[["distinct_n"]], 0)
+    if ("distinct_n" %in% names(sums)) {
+
+      merge_vars <- names(sums)[!(names(sums) %in% c('n', 'distinct_n'))]
+      dist_tot <- sums %>%
+        select(everything(), -n, distinct_total = distinct_n)
+
+      # summary_var may be used for grouping denoms so only toss it if
+      # it's not in denoms_by
+      if (!('summary_var' %in% map_chr(vars_in_denoms, as_name)) & 'summary_var' %in% names(sums)) {
+        merge_vars <- merge_vars[merge_vars != 'summary_var']
+        dist_tot <- dist_tot %>%
+          select(-summary_var)
+      }
+
+      dist_tot <- dist_tot %>% distinct()
+
+      .data <- .data %>%
+        left_join(
+          dist_tot, by = merge_vars
+        )
+
+      if (length(vars_in_denoms) == 1) {
+        # If just treatment group then we can assume should just be filled (i.e.
+        # one denom and no other cols to consider)
+        .data <- .data %>% fill(distinct_total, .direction='downup')
```

Comment:
Can you explain what this fill is doing here? I can't visualize how .data would end up with missing data in the distinct_total column.

Reply:
Well... I guess it's not necessary. 

I went through probably 15 iterations trying to get this sections to work and get all of the unit tests to pass. So at one point, this was necessary, but something else must have resolved the need for it because I removed it and devtools::check() is still passing fine.

I think this entire sections needs a total refactor though. But I'm not sure if I'd want to do that here or just fix it in a successor. As more situations presented, the logic got more and more complicated. The core issue that I see is that originally, get_denom_total basically returned an atomic value - but depending on the denoms that's not true, and it's actually a merge. So instead of the `do()` [here](https://github.com/atorus-research/Tplyr/blob/c31a8e2b7d28e03b52e73b6e83988840a93eed8f/R/count.R#L193) I think we just need the appropriate data frame that merges back to the summary_stat data. 

Double check this section yourself and make sure you're confident in it, But R CMD check is passing for me, and [this](https://github.com/atorus-research/Tplyr/blob/c31a8e2b7d28e03b52e73b6e83988840a93eed8f/tests/testthat/test-count.R#L896) was added specifically to check #136 so I think it's ok to remove this block.

Reply:
Definitly agreement about the refactor, but I think we should refactor without adding any extra functionality to try to keep the complexity of a refactor down. I've been thinking more about how we can improve the workflow to improve the confusion of how the counts and denoms are working.

Reply:
Agreed - please no new functionality...